core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -207,13 +243,155 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
         return topicPartitionData;
     }
 
-    private void releasePartitionLocks(String groupId, Set<TopicIdPartition> topicIdPartitions) {
-        topicIdPartitions.forEach(tp -> {
-            SharePartition sharePartition = sharePartitionManager.sharePartition(groupId, tp);
+    private Map<TopicIdPartition, LogReadResult> maybeReadFromLog(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData) {
+        Map<TopicIdPartition, FetchRequest.PartitionData> missingFetchOffsetMetadataTopicPartitions = new LinkedHashMap<>();
+        topicPartitionData.forEach((topicIdPartition, partitionData) -> {
+            SharePartition sharePartition = sharePartitions.get(topicIdPartition);
+            if (sharePartition.fetchOffsetMetadata().isEmpty()) {
+                missingFetchOffsetMetadataTopicPartitions.put(topicIdPartition, partitionData);
+            }
+        });
+        if (missingFetchOffsetMetadataTopicPartitions.isEmpty()) {
```

Comment:
So did we not chose to implement https://github.com/apache/kafka/pull/17539/files#r1816530978 rather initialize with LinkedHashMap which will hardly be filled?

Reply:
Now, we reset fetchOffsetMetadata everytime the `endOffset` changes (see the usages of function `updateEndOffsetAndResetFetchOffsetMetadata` in `SharePartition`). So, it can be frequent now that the fetchOffsetMetadata is empty. https://github.com/apache/kafka/pull/17539#discussion_r1824931761